### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.loggers.mlflow`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to `pytorch_lightning.loggers.mlflow` for issue #13445 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/loggers/mlflow.py [hunk 1]
```diff
@@ -194,7 +204,7 @@ def run_id(self) -> str:
             The run id.
         """
         _ = self.experiment
-        return self._run_id
+        return self._run_id  # type: ignore[return-value]
```
Comment (otaj) [inline src/pytorch_lightning/loggers/mlflow.py:207]: Can we not ignore types here? Since we are creating the experiment, we know `self._run_id` is not going to be `None` anymore so we can assert against it

File: src/pytorch_lightning/loggers/mlflow.py [hunk 2]
```diff
@@ -204,7 +214,7 @@ def experiment_id(self) -> str:
             The experiment id.
         """
         _ = self.experiment
-        return self._experiment_id
+        return self._experiment_id  # type: ignore[return-value]
```
Comment (otaj) [inline src/pytorch_lightning/loggers/mlflow.py:217]: Same comment applies here

**Review by awaelchli — APPROVED** [review]
after adding assertions LGTM

File: src/pytorch_lightning/loggers/mlflow.py (hunk 2 above)
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/mlflow.py:217]: pyproject.toml can be updated to include [`strict_optional = "False"`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-strict_optional) under [tool.mypy]; which solves this issue where mypy seems to be checking both options ... however, updating the config causes additional errors in other files. 

```sh
src/pytorch_lightning/trainer/connectors/signal_connector.py:153: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:589: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:667: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:669: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:682: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:699: error: Unused "type: ignore" comment
src/pytorch_lightning/trainer/connectors/accelerator_connector.py:755: error: Unused "type: ignore" comment
src/pytorch_lightning/strategies/single_hpu.py:65: error: Unused "type: ignore" comment
src/pytorch_lightning/strategies/fully_sharded_native.py:193: error: Unused "type: ignore" comment
src/pytorch_lightning/strategies/bagua.py:148: error: Unused "type: ignore" comment
src/pytorch_lightning/strategies/bagua.py:149: error: Unused "type: ignore" comment
src/pytorch_lightning/strategies/launchers/spawn.py:118: error: Unused "type: ignore" comment
```

Comment (carmocca) [reply src/pytorch_lightning/loggers/mlflow.py:217]: I don't understand why that would help. We want to be strict here.

File: src/pytorch_lightning/loggers/mlflow.py [hunk 3]
```diff
@@ -250,7 +260,7 @@ def finalize(self, status: str = "FINISHED") -> None:
             self.experiment.set_terminated(self.run_id, status)
 
     @property
-    def save_dir(self) -> Optional[str]:
+    def save_dir(self) -> str:
```
Comment (carmocca) [inline src/pytorch_lightning/loggers/mlflow.py:263]: Why was this changed? It returns `None` if `not self._tracking_uri.startswith(LOCAL_FILE_URI_PREFIX)`

**Review by carmocca — CHANGES_REQUESTED** [review]

File: src/pytorch_lightning/loggers/mlflow.py (hunk 3 above)
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/mlflow.py:263]: self._tracking_uri gets set to a str on init if not provided by the user, with the following at ln 138:

```sh
if not tracking_uri:
    tracking_uri = f"{LOCAL_FILE_URI_PREFIX}{save_dir}"
```

tracking_uri is optional, but it's seem that save_dir can always expect to return a str

File: src/pytorch_lightning/loggers/mlflow.py (hunk 2 above)
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/mlflow.py]: please check latest commit

File: src/pytorch_lightning/loggers/logger.py [hunk 4]
```diff
@@ -203,12 +203,12 @@ def group_separator(self) -> str:
...
     @property
     @abstractmethod
-    def name(self) -> str:
+    def name(self) -> Optional[str]:
         """Return the experiment name."""
 
     @property
     @abstractmethod
-    def version(self) -> Union[int, str]:
+    def version(self) -> Union[int, Optional[str]]:
```
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/logger.py:211]: reverted to Optional[str]

File: src/pytorch_lightning/loggers/mlflow.py [hunk 5]
```diff
@@ -197,7 +207,7 @@ def run_id(self) -> str:
         return self._run_id
 
     @property
-    def experiment_id(self) -> str:
+    def experiment_id(self) -> Optional[str]:
```
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/mlflow.py:210]: reverted to Optional[str]

File: src/pytorch_lightning/loggers/mlflow.py (hunk 3 above)
Comment (carmocca) [reply src/pytorch_lightning/loggers/mlflow.py:263]: But if a user creates a `MLFlowLogger(tracking_uri="foobar")` then the `startswith` will resolve to False and `save_dir` will return `None`. It expects to return a `str` if `tracking_uri` was not provided or was provided as expected.

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/mlflow.py:263]: and the arg save_dir is set to default str; should this have been None?

`save_dir: Optional[str] = "./mlruns"`

**Review by otaj — APPROVED** [review]

**Comment by otaj** [comment]
@carmocca are you fine with the PR now?

File: src/pytorch_lightning/loggers/logger.py [hunk 6]
```diff
@@ -203,12 +203,12 @@ def group_separator(self) -> str:
...
     @property
     @abstractmethod
-    def name(self) -> str:
+    def name(self) -> Optional[str]:
         """Return the experiment name."""
 
     @property
     @abstractmethod
-    def version(self) -> Union[int, str]:
+    def version(self) -> Union[int, str, None]:
```
Comment (carmocca) [inline src/pytorch_lightning/loggers/logger.py:211]:
```suggestion
    def version(self) -> Optional[Union[int, str]]:
```

**Review by carmocca — APPROVED** [review]
